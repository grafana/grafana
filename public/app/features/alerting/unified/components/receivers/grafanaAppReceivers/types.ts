import { SupportedPlugin } from '../../../types/pluginBridges';

export const GRAFANA_APP_RECEIVERS_SOURCE_IMAGE: Record<SupportedPlugin, string> = {
  [SupportedPlugin.OnCall]: 'public/img/alerting/oncall_logo.svg',

  [SupportedPlugin.Incident]: '',
  [SupportedPlugin.MachineLearning]: '',
  [SupportedPlugin.Labels]: '',
  [SupportedPlugin.Slo]: '',
};
