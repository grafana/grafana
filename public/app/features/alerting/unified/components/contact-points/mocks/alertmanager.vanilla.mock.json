{
  "cluster": {
    "name": "01HGAVE6N6RG55BZW8N5ZJ12T3",
    "peers": [{ "address": "172.19.0.6:9094", "name": "01HGAVE6N6RG55BZW8N5ZJ12T3" }],
    "status": "settling"
  },
  "config": {
    "global": {
      "resolve_timeout": "5m",
      "http_config": {
        "tls_config": { "insecure_skip_verify": false },
        "follow_redirects": true,
        "enable_http2": true,
        "proxy_url": null
      },
      "smtp_hello": "localhost",
      "smtp_smarthost": "",
      "smtp_require_tls": true,
      "pagerduty_url": "https://events.pagerduty.com/v2/enqueue",
      "opsgenie_api_url": "https://api.opsgenie.com/",
      "wechat_api_url": "https://qyapi.weixin.qq.com/cgi-bin/",
      "victorops_api_url": "https://alert.victorops.com/integrations/generic/20131114/alert/",
      "telegram_api_url": "https://api.telegram.org",
      "webex_api_url": "https://webexapis.com/v1/messages"
    },
    "route": {
      "receiver": "web.hook",
      "group_by": ["alertname"],
      "group_wait": "30s",
      "group_interval": "5m",
      "repeat_interval": "1h"
    },
    "inhibit_rules": [
      {
        "source_match": { "severity": "critical" },
        "target_match": { "severity": "warning" },
        "equal": ["alertname", "dev", "instance"]
      }
    ],
    "templates": [],
    "receivers": [
      {
        "name": "web.hook",
        "webhook_configs": [
          {
            "send_resolved": true,
            "http_config": {
              "tls_config": { "insecure_skip_verify": false },
              "follow_redirects": true,
              "enable_http2": true,
              "proxy_url": null
            },
            "url": "\u003csecret\u003e",
            "url_file": "",
            "max_alerts": 0
          }
        ]
      }
    ]
  },
  "uptime": "2023-11-28T11:36:10.663Z",
  "versionInfo": {
    "branch": "HEAD",
    "buildDate": "20230824-11:09:02",
    "buildUser": "root@520df6c16a84",
    "goVersion": "go1.20.7",
    "revision": "d7b4f0c7322e7151d6e3b1e31cbc15361e295d8d",
    "version": "0.26.0"
  }
}
