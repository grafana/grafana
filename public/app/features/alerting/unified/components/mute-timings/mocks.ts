import { AlertManagerCortexConfig, MuteTimeInterval } from 'app/plugins/datasource/alertmanager/types';

export const muteTimeInterval: MuteTimeInterval = {
  name: 'default-mute',
  time_intervals: [
    {
      times: [
        {
          start_time: '12:00',
          end_time: '24:00',
        },
      ],
      days_of_month: ['15', '-1'],
      months: ['august:december', 'march'],
    },
  ],
};
/** Alertmanager config where time intervals are stored in `mute_time_intervals` property */
export const defaultConfig: AlertManagerCortexConfig = {
  alertmanager_config: {
    receivers: [{ name: 'default' }, { name: 'critical' }],
    route: {
      receiver: 'default',
      group_by: ['alertname'],
      routes: [
        {
          matchers: ['env=prod', 'region!=EU'],
          mute_time_intervals: [muteTimeInterval.name],
        },
      ],
    },
    templates: [],
    mute_time_intervals: [muteTimeInterval],
  },
  template_files: {},
};
