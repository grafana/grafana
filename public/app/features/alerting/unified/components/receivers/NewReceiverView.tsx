import { useAlertmanager } from 'app/features/alerting/unified/state/AlertmanagerContext';

import { GRAFANA_RULES_SOURCE_NAME } from '../../utils/datasource';

import { CloudReceiverForm } from './form/CloudReceiverForm';
import { GrafanaReceiverForm } from './form/GrafanaReceiverForm';

const NewReceiverView = () => {
  const { selectedAlertmanager } = useAlertmanager();
  if (selectedAlertmanager === GRAFANA_RULES_SOURCE_NAME) {
    return <GrafanaReceiverForm />;
  } else {
    return <CloudReceiverForm alertManagerSourceName={selectedAlertmanager!} />;
  }
};

export default NewReceiverView;
