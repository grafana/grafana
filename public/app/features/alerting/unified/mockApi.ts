import { http, HttpResponse } from 'msw';
import { setupServer, SetupServer } from 'msw/node';

import { DataSourceInstanceSettings } from '@grafana/data';
import { setBackendSrv } from '@grafana/runtime';
import { AlertGroupUpdated } from 'app/features/alerting/unified/api/alertRuleApi';
import allHandlers from 'app/features/alerting/unified/mocks/server/all-handlers';
import { DashboardDTO, FolderDTO, OrgUser } from 'app/types';
import {
  PromBuildInfoResponse,
  PromRulesResponse,
  RulerGrafanaRuleDTO,
  RulerRuleGroupDTO,
  RulerRulesConfigDTO,
} from 'app/types/unified-alerting-dto';

import { backendSrv } from '../../../core/services/backend_srv';
import {
  AlertmanagerConfig,
  AlertManagerCortexConfig,
  AlertmanagerReceiver,
  EmailConfig,
  GrafanaManagedReceiverConfig,
  MatcherOperator,
  Route,
} from '../../../plugins/datasource/alertmanager/types';
import { DashboardSearchItem } from '../../search/types';

type Configurator<T> = (builder: T) => T;

export class AlertmanagerConfigBuilder {
  private alertmanagerConfig: AlertmanagerConfig = { receivers: [] };

  addReceivers(configure: (builder: AlertmanagerReceiverBuilder) => void): AlertmanagerConfigBuilder {
    const receiverBuilder = new AlertmanagerReceiverBuilder();
    configure(receiverBuilder);
    this.alertmanagerConfig.receivers?.push(receiverBuilder.build());
    return this;
  }

  withRoute(configure: (routeBuilder: AlertmanagerRouteBuilder) => void): AlertmanagerConfigBuilder {
    const routeBuilder = new AlertmanagerRouteBuilder();
    configure(routeBuilder);

    this.alertmanagerConfig.route = routeBuilder.build();

    return this;
  }

  build() {
    return this.alertmanagerConfig;
  }
}

class AlertmanagerRouteBuilder {
  private route: Route = { routes: [], object_matchers: [] };

  withReceiver(receiver: string): AlertmanagerRouteBuilder {
    this.route.receiver = receiver;
    return this;
  }

  withoutReceiver(): AlertmanagerRouteBuilder {
    return this;
  }

  withEmptyReceiver(): AlertmanagerRouteBuilder {
    this.route.receiver = '';
    return this;
  }

  addRoute(configure: (builder: AlertmanagerRouteBuilder) => void): AlertmanagerRouteBuilder {
    const routeBuilder = new AlertmanagerRouteBuilder();
    configure(routeBuilder);
    this.route.routes?.push(routeBuilder.build());
    return this;
  }

  addMatcher(key: string, operator: MatcherOperator, value: string): AlertmanagerRouteBuilder {
    this.route.object_matchers?.push([key, operator, value]);
    return this;
  }

  build() {
    return this.route;
  }
}

class EmailConfigBuilder {
  private emailConfig: EmailConfig = { to: '' };

  withTo(to: string): EmailConfigBuilder {
    this.emailConfig.to = to;
    return this;
  }

  build() {
    return this.emailConfig;
  }
}

class GrafanaReceiverConfigBuilder {
  private grafanaReceiverConfig: GrafanaManagedReceiverConfig = {
    name: '',
    type: '',
    settings: {},
    disableResolveMessage: false,
  };

  withType(type: string): GrafanaReceiverConfigBuilder {
    this.grafanaReceiverConfig.type = type;
    return this;
  }

  withName(name: string): GrafanaReceiverConfigBuilder {
    this.grafanaReceiverConfig.name = name;
    return this;
  }

  addSetting(key: string, value: string): GrafanaReceiverConfigBuilder {
    if (this.grafanaReceiverConfig.settings) {
      this.grafanaReceiverConfig.settings[key] = value;
    }
    return this;
  }

  build() {
    return this.grafanaReceiverConfig;
  }
}

export class AlertmanagerReceiverBuilder {
  private receiver: AlertmanagerReceiver = { name: '', email_configs: [], grafana_managed_receiver_configs: [] };

  withName(name: string): AlertmanagerReceiverBuilder {
    this.receiver.name = name;
    return this;
  }

  addGrafanaReceiverConfig(configure: Configurator<GrafanaReceiverConfigBuilder>): AlertmanagerReceiverBuilder {
    this.receiver.grafana_managed_receiver_configs?.push(configure(new GrafanaReceiverConfigBuilder()).build());
    return this;
  }

  addEmailConfig(configure: (builder: EmailConfigBuilder) => void): AlertmanagerReceiverBuilder {
    const builder = new EmailConfigBuilder();
    configure(builder);
    this.receiver.email_configs?.push(builder.build());
    return this;
  }

  build() {
    return this.receiver;
  }
}

export function mockApi(server: SetupServer) {
  return {
    getAlertmanagerConfig: (amName: string, configure: (builder: AlertmanagerConfigBuilder) => void) => {
      const builder = new AlertmanagerConfigBuilder();
      configure(builder);

      server.use(
        http.get(`api/alertmanager/${amName}/config/api/v1/alerts`, () =>
          HttpResponse.json<AlertManagerCortexConfig>({
            alertmanager_config: builder.build(),
            template_files: {},
          })
        )
      );
    },
  };
}

export function mockAlertRuleApi(server: SetupServer) {
  return {
    prometheusRuleNamespaces: (dsName: string, response: PromRulesResponse) => {
      server.use(
        http.get(`api/prometheus/${dsName}/api/v1/rules`, () => HttpResponse.json<PromRulesResponse>(response))
      );
    },
    rulerRules: (dsName: string, response: RulerRulesConfigDTO) => {
      server.use(http.get(`/api/ruler/${dsName}/api/v1/rules`, () => HttpResponse.json(response)));
    },
    updateRule: (dsName: string, response: AlertGroupUpdated) => {
      server.use(http.post(`/api/ruler/${dsName}/api/v1/rules/:namespaceUid`, () => HttpResponse.json(response)));
    },
    rulerRuleGroup: (dsName: string, namespace: string, group: string, response: RulerRuleGroupDTO) => {
      server.use(
        http.get(`/api/ruler/${dsName}/api/v1/rules/${namespace}/${group}`, () => HttpResponse.json(response))
      );
    },
    getAlertRule: (uid: string, response: RulerGrafanaRuleDTO) => {
      server.use(http.get(`/api/ruler/grafana/api/v1/rule/${uid}`, () => HttpResponse.json(response)));
    },
  };
}

/**
 * Used to mock the response from the /api/v1/status/buildinfo endpoint
 */
export function mockFeatureDiscoveryApi(server: SetupServer) {
  return {
    /**
     *
     * @param dsSettings Use `mockDataSource` to create a faks data source settings
     * @param response Use `buildInfoResponse` to get a pre-defined response for Prometheus and Mimir
     */
    discoverDsFeatures: (dsSettings: DataSourceInstanceSettings, response: PromBuildInfoResponse) => {
      server.use(http.get(`${dsSettings.url}/api/v1/status/buildinfo`, () => HttpResponse.json(response)));
    },
  };
}

export function mockExportApi(server: SetupServer) {
  // exportRule, exportRulesGroup, exportRulesFolder use the same API endpoint but with different parameters
  return {
    // exportRulesGroup requires folderUid and group parameters and doesn't allow ruleUid parameter
    exportRulesGroup: (folderUid: string, group: string, response: Record<string, string>) => {
      server.use(
        http.get('/api/ruler/grafana/api/v1/export/rules', ({ request }) => {
          const url = new URL(request.url);
          if (url.searchParams.get('folderUid') === folderUid && url.searchParams.get('group') === group) {
            const format = url.searchParams.get('format') ?? 'yaml';
            return HttpResponse.text(response[format]);
          }

          return HttpResponse.text('', { status: 500 });
        })
      );
    },
    modifiedExport: (namespaceUID: string, response: Record<string, string>) => {
      server.use(
        http.post(`/api/ruler/grafana/api/v1/rules/${namespaceUID}/export`, ({ request }) => {
          const url = new URL(request.url);
          const format = url.searchParams.get('format') ?? 'yaml';
          return HttpResponse.text(response[format]);
        })
      );
    },
  };
}

export function mockFolderApi(server: SetupServer) {
  return {
    folder: (folderUid: string, response: FolderDTO) => {
      server.use(http.get(`/api/folders/${folderUid}`, () => HttpResponse.json(response)));
    },
  };
}

export function mockSearchApi(server: SetupServer) {
  return {
    search: (results: DashboardSearchItem[]) => {
      server.use(http.get(`/api/search`, () => HttpResponse.json(results)));
    },
  };
}

export function mockUserApi(server: SetupServer) {
  return {
    user: (user: OrgUser) => {
      server.use(http.get(`/api/user`, () => HttpResponse.json(user)));
    },
  };
}

export function mockDashboardApi(server: SetupServer) {
  return {
    search: (results: DashboardSearchItem[]) => {
      server.use(http.get(`/api/search`, () => HttpResponse.json(results)));
    },
    dashboard: (response: DashboardDTO) => {
      server.use(http.get(`/api/dashboards/uid/${response.dashboard.uid}`, () => HttpResponse.json(response)));
    },
  };
}

const server = setupServer(...allHandlers);

/**
 * Sets up beforeAll, afterAll and beforeEach handlers for mock server
 */
export function setupMswServer() {
  beforeAll(() => {
    setBackendSrv(backendSrv);
    server.listen({ onUnhandledRequest: 'error' });
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  return server;
}

export default server;
