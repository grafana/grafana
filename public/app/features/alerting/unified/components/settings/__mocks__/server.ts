import { delay, http, HttpResponse } from 'msw';
import { SetupServerApi } from 'msw/node';

import { setDataSourceSrv } from '@grafana/runtime';
import {
  AlertManagerCortexConfig,
  AlertManagerDataSourceJsonData,
  AlertManagerImplementation,
  AlertmanagerReceiver,
  Receiver,
} from 'app/plugins/datasource/alertmanager/types';

import { mockDataSource, MockDataSourceSrv } from '../../../mocks';
import * as config from '../../../utils/config';
import { DataSourceType } from '../../../utils/datasource';

import internalAlertmanagerConfig from './api/alertmanager/grafana/config/api/v1/alerts.json';
import history from './api/alertmanager/grafana/config/history.json';
import cloudAlertmanagerConfig from './api/alertmanager/provisioned/config/api/v1/alerts.json';
import vanillaAlertmanagerConfig from './api/alertmanager/vanilla prometheus/api/v2/status.json';
import datasources from './api/datasources.json';
import admin_config from './api/v1/ngalert/admin_config.json';
import alertmanagers from './api/v1/ngalert/alertmanagers.json';

export { datasources as DataSourcesResponse };
export { admin_config as AdminConfigResponse };
export { alertmanagers as AlertmanagersResponse };
export { internalAlertmanagerConfig as InternalAlertmanagerConfiguration };
export { vanillaAlertmanagerConfig as VanillaAlertmanagerConfiguration };
export { history as alertmanagerConfigurationHistory };

export const EXTERNAL_VANILLA_ALERTMANAGER_UID = 'vanilla-alertmanager';
export const PROVISIONED_MIMIR_ALERTMANAGER_UID = 'provisioned-alertmanager';

jest.spyOn(config, 'getAllDataSources');

const mocks = {
  getAllDataSources: jest.mocked(config.getAllDataSources),
};

export const mockDataSources = {
  [EXTERNAL_VANILLA_ALERTMANAGER_UID]: mockDataSource<AlertManagerDataSourceJsonData>({
    uid: EXTERNAL_VANILLA_ALERTMANAGER_UID,
    name: EXTERNAL_VANILLA_ALERTMANAGER_UID,
    type: DataSourceType.Alertmanager,
    jsonData: {
      implementation: AlertManagerImplementation.prometheus,
    },
  }),
  [PROVISIONED_MIMIR_ALERTMANAGER_UID]: mockDataSource<AlertManagerDataSourceJsonData>({
    uid: PROVISIONED_MIMIR_ALERTMANAGER_UID,
    name: PROVISIONED_MIMIR_ALERTMANAGER_UID,
    type: DataSourceType.Alertmanager,
    jsonData: {
      // this is a mutable data source type but we're making it readOnly
      implementation: AlertManagerImplementation.mimir,
    },
    readOnly: true,
  }),
};

export function setupGrafanaManagedServer(server: SetupServerApi) {
  server.use(
    createAdminConfigHandler(),
    createExternalAlertmanagersHandler(),
    createAlertmanagerDataSourcesHandler(),
    ...createAlertmanagerConfigurationHandlers(),
    createAlertmanagerHistoryHandler()
  );

  return server;
}

export function setupVanillaAlertmanagerServer(server: SetupServerApi) {
  mocks.getAllDataSources.mockReturnValue(Object.values(mockDataSources));
  setDataSourceSrv(new MockDataSourceSrv(mockDataSources));

  server.use(
    createVanillaAlertmanagerConfigurationHandler(EXTERNAL_VANILLA_ALERTMANAGER_UID),
    ...createAlertmanagerConfigurationHandlers()
  );

  return server;
}

const createAdminConfigHandler = () => http.get('/api/v1/ngalert/admin_config', () => HttpResponse.json(admin_config));

const createExternalAlertmanagersHandler = () => {
  return http.get('/api/v1/ngalert/alertmanagers', () => HttpResponse.json(alertmanagers));
};

const createAlertmanagerConfigurationHandlers = () => {
  // Dirty check to type guard against us having a non-Grafana managed receiver
  const contactPointIsAMReceiver = (receiver: Receiver): receiver is AlertmanagerReceiver => {
    return !receiver.grafana_managed_receiver_configs;
  };

  return [
    http.get<{ name: string }>(`/api/alertmanager/:name/config/api/v1/alerts`, ({ params }) => {
      if (params.name === 'grafana') {
        return HttpResponse.json(internalAlertmanagerConfig);
      }
      return HttpResponse.json(cloudAlertmanagerConfig);
    }),
    http.post<never, AlertManagerCortexConfig>(`/api/alertmanager/:name/config/api/v1/alerts`, async ({ request }) => {
      await delay(1000); // simulate some time

      // Specifically mock and check for the case of an invalid telegram config,
      // and return a 400 error in this case
      // This is to test against us accidentally sending a `{label, value}` object instead of a string
      const body = await request.json();
      const invalidConfig = body.alertmanager_config.receivers?.some((receiver) => {
        if (!contactPointIsAMReceiver(receiver)) {
          return false;
        }

        const invalidParseMode = (receiver.telegram_configs || []).some(
          (config) => typeof config.parse_mode === 'object'
        );
        const invalidChatId = (receiver.telegram_configs || []).some((config) => Number(config.chat_id) >= 0);

        return invalidParseMode || invalidChatId;
      });

      if (invalidConfig) {
        return HttpResponse.json({ message: 'bad request data' }, { status: 400 });
      }

      return HttpResponse.json({ message: 'configuration created' });
    }),
  ];
};

const createAlertmanagerDataSourcesHandler = () => http.get('/api/datasources', () => HttpResponse.json(datasources));
const createAlertmanagerHistoryHandler = (name = 'grafana') =>
  http.get(`/api/alertmanager/${name}/config/history`, () => HttpResponse.json(history));

const createVanillaAlertmanagerConfigurationHandler = (dataSourceUID: string) =>
  http.get(`/api/alertmanager/${dataSourceUID}/api/v2/status`, () => HttpResponse.json(vanillaAlertmanagerConfig));

export const withExternalOnlySetting = (server: SetupServerApi) => {
  server.use(createAdminConfigHandler());
};
