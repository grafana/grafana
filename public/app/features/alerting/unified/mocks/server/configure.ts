import { HttpResponse, http } from 'msw';

import { config } from '@grafana/runtime';
import server, { mockFeatureDiscoveryApi } from 'app/features/alerting/unified/mockApi';
import { mockDataSource, mockFolder } from 'app/features/alerting/unified/mocks';
import {
  ALERTMANAGER_UPDATE_ERROR_RESPONSE,
  getAlertmanagerConfigHandler,
  getGrafanaAlertmanagerConfigHandler,
  grafanaAlertingConfigurationStatusHandler,
  updateGrafanaAlertmanagerConfigHandler,
} from 'app/features/alerting/unified/mocks/server/handlers/alertmanagers';
import { getFolderHandler } from 'app/features/alerting/unified/mocks/server/handlers/folders';
import { listNamespacedTimeIntervalHandler } from 'app/features/alerting/unified/mocks/server/handlers/k8s/timeIntervals.k8s';
import {
  getDisabledPluginHandler,
  getPluginMissingHandler,
} from 'app/features/alerting/unified/mocks/server/handlers/plugins';
import { SupportedPlugin } from 'app/features/alerting/unified/types/pluginBridges';
import { clearPluginSettingsCache } from 'app/features/plugins/pluginSettings';
import { AlertManagerCortexConfig, AlertmanagerChoice } from 'app/plugins/datasource/alertmanager/types';
import { FolderDTO } from 'app/types';

import { setupDataSources } from '../../testSetup/datasources';
import { buildInfoResponse } from '../../testSetup/featureDiscovery';
import { DataSourceType } from '../../utils/datasource';

import { MIMIR_DATASOURCE_UID } from './constants';
import { rulerRuleGroupHandler, updateRulerRuleNamespaceHandler } from './handlers/grafanaRuler';

export type HandlerOptions = {
  delay?: number;
  response?: HttpResponse;
};

/**
 * Makes the mock server respond in a way that matches the different behaviour associated with
 * Alertmanager choices and the number of configured external alertmanagers
 */
export const setAlertmanagerChoices = (alertmanagersChoice: AlertmanagerChoice, numExternalAlertmanagers: number) => {
  const response = {
    alertmanagersChoice,
    numExternalAlertmanagers,
  };
  server.use(grafanaAlertingConfigurationStatusHandler(response));
};

/**
 * Makes the mock server respond with different folder access control settings
 */
export const setFolderAccessControl = (accessControl: FolderDTO['accessControl']) => {
  server.use(getFolderHandler(mockFolder({ hasAcl: true, accessControl })));
};

/**
 * Makes the mock server respond with different folder response, for just the folder in question
 */
export const setFolderResponse = (response: Partial<FolderDTO>) => {
  const handler = http.get<{ folderUid: string }>(`/api/folders/${response.uid}`, () => HttpResponse.json(response));
  server.use(handler);
};

/**
 * Makes the mock server respond with different Grafana Alertmanager config
 */
export const setGrafanaAlertmanagerConfig = (config: AlertManagerCortexConfig) => {
  server.use(getGrafanaAlertmanagerConfigHandler(config));
};

/**
 * Makes the mock server respond with different (other) Alertmanager config
 */
export const setAlertmanagerConfig = (config: AlertManagerCortexConfig) => {
  server.use(getAlertmanagerConfigHandler(config));
};

/**
 * Makes the mock server respond with different responses for updating a ruler namespace
 */
export const setUpdateRulerRuleNamespaceHandler = (options?: HandlerOptions) => {
  const handler = updateRulerRuleNamespaceHandler(options);
  server.use(handler);

  return handler;
};

/**
 * Makes the mock server respond with different responses for a ruler rule group
 */
export const setRulerRuleGroupHandler = (options?: HandlerOptions) => {
  const handler = rulerRuleGroupHandler(options);
  server.use(handler);

  return handler;
};

/**
 * Makes the mock server respond with an error when fetching list of mute timings
 */
export const setMuteTimingsListError = () => {
  const listMuteTimingsPath = listNamespacedTimeIntervalHandler().info.path;
  const handler = http.get(listMuteTimingsPath, () => {
    return HttpResponse.json({}, { status: 401 });
  });

  server.use(handler);
  return handler;
};

export function mimirDataSource() {
  const dataSource = mockDataSource(
    {
      type: DataSourceType.Prometheus,
      name: MIMIR_DATASOURCE_UID,
      uid: MIMIR_DATASOURCE_UID,
      url: 'https://mimir.local:9000',
      jsonData: {
        manageAlerts: true,
      },
    },
    { alerting: true }
  );

  setupDataSources(dataSource);
  mockFeatureDiscoveryApi(server).discoverDsFeatures(dataSource, buildInfoResponse.mimir);

  return { dataSource };
}

/** Make a given plugin ID respond with a 404, as if it isn't installed at all */
export const removePlugin = (pluginId: string) => {
  delete config.apps[pluginId];
  server.use(getPluginMissingHandler(pluginId));
};

/** Make a plugin respond with `enabled: false`, as if its installed but disabled */
export const disablePlugin = (pluginId: SupportedPlugin) => {
  clearPluginSettingsCache(pluginId);
  server.use(getDisabledPluginHandler(pluginId));
};

/** Make alertmanager config update fail */
export const makeGrafanaAlertmanagerConfigUpdateFail = () => {
  server.use(updateGrafanaAlertmanagerConfigHandler(ALERTMANAGER_UPDATE_ERROR_RESPONSE));
};
