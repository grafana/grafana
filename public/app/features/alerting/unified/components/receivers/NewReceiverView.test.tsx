import { Routes, Route } from 'react-router-dom-v5-compat';
import { render, screen } from 'test/test-utils';
import { byLabelText, byPlaceholderText, byRole, byTestId } from 'testing-library-selector';

import { makeAlertmanagerConfigUpdateFail } from 'app/features/alerting/unified/mocks/server/configure';
import { captureRequests } from 'app/features/alerting/unified/mocks/server/events';
import { AccessControlAction } from 'app/types';

import { setupMswServer } from '../../mockApi';
import { grantUserPermissions } from '../../mocks';
import { testWithFeatureToggles } from '../../test/test-utils';

import NewReceiverView from './NewReceiverView';

setupMswServer();

const Index = () => {
  return <div>redirected</div>;
};

const renderForm = () =>
  render(
    <Routes>
      <Route path="/alerting/notifications" element={<Index />} />
      <Route path="/alerting/notifications/new" element={<NewReceiverView />} />
    </Routes>,
    {
      historyOptions: { initialEntries: ['/alerting/notifications/new'] },
    }
  );

beforeEach(() => {
  grantUserPermissions([AccessControlAction.AlertingNotificationsRead, AccessControlAction.AlertingNotificationsWrite]);
});

describe('alerting API server enabled', () => {
  testWithFeatureToggles(['alertingApiServer']);

  it('can create a receiver', async () => {
    const { user } = renderForm();

    expect(await ui.inputs.name.find()).toBeEnabled();

    await user.type(await ui.inputs.name.find(), 'my new receiver');

    // enter some email
    const email = ui.inputs.email.addresses.get();
    await user.clear(email);
    await user.type(email, 'tester@grafana.com');

    await user.click(ui.saveContactButton.get());

    expect(await screen.findByText(/redirected/i)).toBeInTheDocument();
  });
});

describe('alerting API server disabled', () => {
  it('should be able to test and save a receiver', async () => {
    const capture = captureRequests();

    const { user } = renderForm();

    // wait for loading to be done
    // type in a name for the new receiver
    await user.type(await ui.inputs.name.find(), 'my new receiver');

    // enter some email
    const email = ui.inputs.email.addresses.get();
    await user.clear(email);
    await user.type(email, 'tester@grafana.com');

    // try to test the contact point
    await user.click(await ui.testContactPointButton.find());

    expect(await ui.testContactPointModal.find()).toBeInTheDocument();

    await user.click(ui.customContactPointOption.get());

    // enter custom annotations and labels
    await user.type(screen.getByPlaceholderText('Enter a description...'), 'Test contact point');
    await user.type(ui.contactPointLabelKey(0).get(), 'foo');
    await user.type(ui.contactPointLabelValue(0).get(), 'bar');

    // click test
    await user.click(ui.testContactPoint.get());

    // we shouldn't be testing implementation details but when the request is successful
    // it can't seem to assert on the success toast
    await user.click(ui.saveContactButton.get());

    const requests = await capture;
    const testRequest = requests.find((r) => r.url.endsWith('/config/api/v1/receivers/test'));
    const saveRequest = requests.find(
      (r) => r.url.endsWith('/api/alertmanager/grafana/config/api/v1/alerts') && r.method === 'POST'
    );

    const testBody = await testRequest?.json();
    const fullSaveBody = await saveRequest?.json();

    // Only snapshot and check the receivers, as we don't want other tests to break this
    // just because we added something new to the mock config
    const saveBody = fullSaveBody.alertmanager_config.receivers;

    expect([testBody]).toMatchSnapshot();
    expect([saveBody]).toMatchSnapshot();
  });

  it('does not redirect when creating contact point and API errors', async () => {
    makeAlertmanagerConfigUpdateFail();
    const { user } = renderForm();

    await user.type(await ui.inputs.name.find(), 'receiver that should fail');
    const email = ui.inputs.email.addresses.get();
    await user.clear(email);
    await user.type(email, 'tester@grafana.com');

    await user.click(ui.saveContactButton.get());

    expect(screen.queryByText(/redirected/i)).not.toBeInTheDocument();
  });
});

const ui = {
  saveContactButton: byRole('button', { name: /save contact point/i }),

  testContactPointButton: byRole('button', { name: /Test/ }),
  testContactPointModal: byRole('heading', { name: /test contact point/i }),
  customContactPointOption: byRole('radio', { name: /custom/i }),
  contactPointLabelKey: (idx: number) => byTestId(`label-key-${idx}`),
  contactPointLabelValue: (idx: number) => byTestId(`label-value-${idx}`),
  testContactPoint: byRole('button', { name: /send test notification/i }),
  inputs: {
    name: byPlaceholderText('Name'),
    email: {
      addresses: byLabelText(/Addresses/),
    },
  },
};
