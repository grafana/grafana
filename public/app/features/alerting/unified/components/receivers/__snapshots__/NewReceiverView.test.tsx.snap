// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alerting API server disabled should be able to test and save a receiver 1`] = `
[
  {
    "alert": {
      "annotations": {
        "description": "Test contact point",
      },
      "labels": {
        "foo": "bar",
      },
    },
    "receivers": [
      {
        "grafana_managed_receiver_configs": [
          {
            "disableResolveMessage": false,
            "name": "test",
            "secureSettings": {},
            "settings": {
              "addresses": "tester@grafana.com",
              "singleEmail": false,
            },
            "type": "email",
          },
        ],
        "name": "test",
      },
    ],
  },
]
`;

exports[`alerting API server disabled should be able to test and save a receiver 2`] = `
[
  [
    {
      "grafana_managed_receiver_configs": [
        {
          "disableResolveMessage": false,
          "name": "grafana-default-email",
          "secureFields": {},
          "settings": {
            "addresses": "gilles.demey@grafana.com",
            "singleEmail": false,
          },
          "type": "email",
          "uid": "xeKQrBrnk",
        },
      ],
      "name": "grafana-default-email",
    },
    {
      "grafana_managed_receiver_configs": [
        {
          "disableResolveMessage": false,
          "name": "provisioned-contact-point",
          "provenance": "api",
          "secureFields": {},
          "settings": {
            "addresses": "gilles.demey@grafana.com",
            "singleEmail": false,
          },
          "type": "email",
          "uid": "s8SdCVjnk",
        },
      ],
      "name": "provisioned-contact-point",
    },
    {
      "grafana_managed_receiver_configs": [
        {
          "disableResolveMessage": false,
          "name": "lotsa-emails",
          "secureFields": {},
          "settings": {
            "addresses": "gilles.demey+1@grafana.com, gilles.demey+2@grafana.com, gilles.demey+3@grafana.com, gilles.demey+4@grafana.com",
            "message": "{{ template "slack-template" . }}",
            "singleEmail": false,
            "subject": "some custom value",
          },
          "type": "email",
          "uid": "af306c96-35a2-4d6e-908a-4993e245dbb2",
        },
      ],
      "name": "lotsa-emails",
    },
    {
      "grafana_managed_receiver_configs": [
        {
          "disableResolveMessage": false,
          "name": "Slack with multiple channels",
          "secureFields": {
            "token": true,
          },
          "settings": {
            "recipient": "test-alerts",
          },
          "type": "slack",
          "uid": "c02ad56a-31da-46b9-becb-4348ec0890fd",
        },
        {
          "disableResolveMessage": false,
          "name": "Slack with multiple channels",
          "secureFields": {
            "token": true,
          },
          "settings": {
            "recipient": "test-alerts2",
          },
          "type": "slack",
          "uid": "b286a3be-f690-49e2-8605-b075cbace2df",
        },
      ],
      "name": "Slack with multiple channels",
    },
    {
      "grafana_managed_receiver_configs": [
        {
          "disableResolveMessage": false,
          "name": "Oncall-integration",
          "settings": {
            "url": "https://oncall-endpoint.example.com",
          },
          "type": "oncall",
        },
      ],
      "name": "OnCall Conctact point",
    },
    {
      "grafana_managed_receiver_configs": [
        {
          "disableResolveMessage": false,
          "name": "my new receiver",
          "secureSettings": {},
          "settings": {
            "addresses": "tester@grafana.com",
            "singleEmail": false,
          },
          "type": "email",
        },
      ],
      "name": "my new receiver",
    },
  ],
]
`;
